### PR TITLE
[cuDNN] Migrate to new frontend API for batchnorm

### PR DESCRIPTION
**Pull request by AaronWang04**
Add v8 cuDNN frontend API for batchnorm

Theoretically new api is a bit faster at cost of higher memory usage, but really this PR is necessary as cuDNN plans to deprecate the old API in upcoming versions.

In order to not break backward compatibility (in case of old cudnn version) it adds a fallback which dispatches to old API if an env flag (TORCH_CUDNN_V8_API_DISABLED) is set. This logic is borrowed from cudnn conv which has the same set of v7/v8 apis.

Follows the example shown here: https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/cpp/norm/batchnorm.cpp

### CODE REVIEW

**Comment by janeyx99** [comment]
What's the benefit and cost of this PR? I'm imagining the cost is increase in binary size/complexity? And the pro should be that...it's faster? It'd be helpful to have a description for why this change is good.

File: aten/src/ATen/native/cudnn/BatchNorm_v8.cpp [hunk 1]
```diff
@@ -0,0 +1,911 @@
...
+
+namespace at {
+namespace native {
+
+namespace {
+namespace fe = cudnn_frontend;
+
+auto get_fe_dtype(const Tensor& t) {
+  auto dtype = t.scalar_type();
+  auto fe_dtype = fe::DataType_t::FLOAT;
```
Comment (Skylion007) [inline aten/src/ATen/native/cudnn/BatchNorm_v8.cpp]: Why specify a default fe_dtype at all? Is it needed for auto I guess? Wouldn't this be a text book example of switch statement handling btw?

File: aten/src/ATen/native/cudnn/BatchNorm_v8.cpp [hunk 2]
```diff
@@ -0,0 +1,911 @@
...
+
+      auto [Y_temp, mean_temp, inv_var_temp, next_mean_temp, next_var_temp] =
+          batchnorm_graph->batchnorm(
+              X_fe, scale_fe, bias_fe, batchnorm_options);
+
+      Y_fe = Y_temp;
+      saved_mean_fe = mean_temp;
+      saved_inv_var_fe = inv_var_temp;
+      next_running_mean_fe = next_mean_temp;
+      next_running_var_fe = next_var_temp;
```
Comment (Skylion007) [inline aten/src/ATen/native/cudnn/BatchNorm_v8.cpp:324]:
```suggestion
      Y_fe = std::move(Y_temp);
      saved_mean_fe = std::move(mean_temp);
      saved_inv_var_fe = std::move(inv_var_temp);
      next_running_mean_fe = std::move(next_mean_temp);
      next_running_var_fe = std::move(next_var_temp);

File: aten/src/ATen/native/cudnn/BatchNorm_v8.cpp [hunk 3]
```diff
@@ -0,0 +1,911 @@
...
+
+    std::unordered_map<std::shared_ptr<fe::graph::Tensor_attributes>, void*>
+        variant_pack_ = {
+            {X_fe, X.data_ptr()},
+            {scale_fe, weight.data_ptr()},
+            {bias_fe, bias.data_ptr()},
+            {Y_fe, Y->data_ptr()}};
+
+    if (training) {
+      variant_pack_[saved_mean_fe] = save_mean->data_ptr();
```
Comment (Skylion007) [inline aten/src/ATen/native/cudnn/BatchNorm_v8.cpp:391]: These should all be const_data_ptr or mutable_data_ptr

File: aten/src/ATen/native/cudnn/BatchNorm_v8.cpp [hunk 4]
```diff
@@ -0,0 +1,911 @@
...
+        X_fe,
+        scale_fe,
+        bias_fe,
+        running_mean_fe,
+        running_var_fe,
+        Y_fe,
+        saved_mean_fe,
+        saved_inv_var_fe,
+        next_running_mean_fe,
+        next_running_var_fe);
```
Comment (Skylion007) [inline aten/src/ATen/native/cudnn/BatchNorm_v8.cpp]:
```suggestion
        std::move(batchnorm_graph),
        std::move(X_fe),
        std::move(scale_fe),
        std::move(bias_fe),
        std::move(running_mean_fe),
        std::move(running_var_fe),
        std::move(Y_fe),
        std::move(saved_mean_fe),
        std::move(saved_inv_var_fe),
        std::move(next_running_mean_fe),
        std::move(next_running_var_fe));

File: aten/src/ATen/native/cudnn/BatchNorm_v8.cpp [hunk 5]
```diff
@@ -0,0 +1,911 @@
...
+    auto result = std::make_tuple(
+        batchnorm_graph,
+        X_fe,
+        DY_fe,
+        scale_fe,
+        mean_fe,
+        inv_variance_fe,
+        DX_fe,
+        dscale_fe,
+        dbias_fe);
```
Comment (Skylion007) [inline aten/src/ATen/native/cudnn/BatchNorm_v8.cpp]:
```suggestion
        std::move(batchnorm_graph),
        std::move(X_fe),
        std::move(DY_fe),
        std::move(scale_fe),
        std::move(mean_fe),
        std::move(inv_variance_fe),
        std::move(DX_fe),
        std::move(dscale_fe),
        std::move(dbias_fe));

File: aten/src/ATen/native/cudnn/BatchNorm_v8.cpp [hunk 6]
```diff
@@ -0,0 +1,911 @@
...
+      *save_var,
+      true, // training mode for backward
+      epsilon,
+      &grad_input_t,
+      &grad_weight_t,
+      &grad_bias_t,
+      input->suggest_memory_format());
+
+  return std::tuple<Tensor, Tensor, Tensor>{
+      grad_input_t, grad_weight_t, grad_bias_t};
```
Comment (Skylion007) [inline aten/src/ATen/native/cudnn/BatchNorm_v8.cpp]:
```suggestion
      std::move(grad_input_t), std::move(grad_weight_t), std::move(grad_bias_t)};

**Review by Skylion007 — COMMENTED** [review]
Looks good to me aside from missing moves causing some unnecessary atomic ref counts increments.

File: aten/src/ATen/native/cudnn/BatchNorm_v8.cpp (hunk 3 above)
Comment (AaronWang04) [reply aten/src/ATen/native/cudnn/BatchNorm_v8.cpp:391]: cuDNN weirdly only accepts void pointers
